### PR TITLE
GlobalDNS Provider and Entry Resources

### PR DESCRIPTION
**Pull request by dahendel**
Adds the following:
- resource: rancher2_global_dns_provider and rancher2_global_dns_provider
- data source: rancher2_global_dns_provider

### CODE REVIEW

File: rancher2/data_source_rancher2_global_dns_provider_test.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  provider = "route53"
+  root_domain = "example.com"
+
+  route53_config = {
+	access_key = "YYYYYYYYYYYYYYYYYYYY"
+	secret_key = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
+	zone_type = "private"
+    region = "us-east-1"
+  }
+}
```
Comment (rawmind0) [inline rancher2/data_source_rancher2_global_dns_provider_test.go:30]: Could you please indent section properly??

```
  route53_config = {
    access_key = "YYYYYYYYYYYYYYYYYYYY"
    secret_key = "XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX"
    zone_type = "private"
    region = "us-east-1"
  }
```

File: rancher2/schema_global_dns_provider.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+				provider := v.(string)
+				for _, p := range validDNSProviders {
+					if p == provider {
+						//strings = append(strings, provider)
+						return
+					}
+				}
+				errors = append(errors, fmt.Errorf("provider must be one of aws, cloudflare, alidns"))
+				return
+			},
```
Comment (rawmind0) [inline rancher2/schema_global_dns_provider.go]: importing `"github.com/hashicorp/terraform-plugin-sdk/helper/validation"` module, you could use this ValidateFunc

```
ValidateFunc: validation.StringInSlice([]string{"route53", "cloudflare", "alidns"}, false),
```

File: rancher2/data_source_rancher2_global_dns_provider_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+data "` + testAccRancher2GlobalDNSProviderDataSourceType + `" "foo" {
+  name = "${rancher2_global_role_binding.foo.name}"
```
Comment (rawmind0) [inline rancher2/data_source_rancher2_global_dns_provider_test.go]: name should be referencing `${rancher2_global_dns_provider.name}`

File: rancher2/data_source_rancher2_global_dns_provider_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	testAccRancher2GlobalDNSProviderDataSourceType = "rancher2_global_global_dns_provider"
+)
+
+var (
+	testAccCheckRancher2GlobalDNSProviderDataSourceConfig string
+)
+
+func init() {
+	testAccCheckRancher2GlobalDNSProviderDataSourceConfig = `
+resource "rancher2_global_dns_config" "foo" {
```
Comment (rawmind0) [inline rancher2/data_source_rancher2_global_dns_provider_test.go]: Resource should be `rancher2_global_dns_provider`

File: rancher2/data_source_rancher2_global_dns_provider_test.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
+package rancher2
+
+import (
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+)
+
+const (
+	testAccRancher2GlobalDNSProviderDataSourceType = "rancher2_global_global_dns_provider"
```
Comment (rawmind0) [inline rancher2/data_source_rancher2_global_dns_provider_test.go]: `testAccRancher2GlobalDNSProviderDataSourceType ` should be `rancher2_global_dns_provider`

File: rancher2/data_source_rancher2_global_dns_provider_test.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  }
+}
+
+data "` + testAccRancher2GlobalDNSProviderDataSourceType + `" "foo" {
+  name = "${rancher2_global_role_binding.foo.name}"
+}
+`
+}
+
+func TestAccRancher2GlobalDNSProviderataSource(t *testing.T) {
```
Comment (rawmind0) [inline rancher2/data_source_rancher2_global_dns_provider_test.go]: Function name should be `TestAccRancher2GlobalDNSProviderDataSource`

File: rancher2/resource_rancher2_global_dns_provider_test.go [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+	resource.Test(t, resource.TestCase{
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckRancher2GlobalDNSProviderDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccRancher2GlobalDNSProviderConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRancher2GlobalDNSProviderExists(testAccRancher2GlobalDNSProviderType+".foo", globalDNSProvider),
+					resource.TestCheckResourceAttr(testAccRancher2GlobalDNSProviderType+".foo", "name", "foo-test"),
+					resource.TestCheckResourceAttr(testAccRancher2GlobalDNSProviderType+".foo", "global_role_id", "user-base"),
```
Comment (rawmind0) [inline rancher2/resource_rancher2_global_dns_provider_test.go]: What is `global_role_id` argument??

File: rancher2/resource_rancher2_global_dns_provider_test.go [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+				),
+			},
+			resource.TestStep{
+				Config: testAccRancher2GlobalDNSProviderUpdateConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRancher2GlobalDNSProviderExists(testAccRancher2GlobalDNSProviderType+".foo", globalDNSProvider),
+					resource.TestCheckResourceAttr(testAccRancher2GlobalDNSProviderType+".foo", "name", "foo-test-updated"),
+					resource.TestCheckResourceAttr(testAccRancher2GlobalDNSProviderType+".foo", "global_role_id", "user-base"),
```
Comment (rawmind0) [inline rancher2/resource_rancher2_global_dns_provider_test.go]: What is `global_role_id` argument??

File: rancher2/resource_rancher2_global_dns_provider_test.go [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+				),
+			},
+			resource.TestStep{
+				Config: testAccRancher2GlobalDNSProviderRecreateConfig,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckRancher2GlobalDNSProviderExists(testAccRancher2GlobalDNSProviderType+".foo", globalDNSProvider),
+					resource.TestCheckResourceAttr(testAccRancher2GlobalDNSProviderType+".foo", "name", "foo-test"),
+					resource.TestCheckResourceAttr(testAccRancher2GlobalDNSProviderType+".foo", "global_role_id", "user-base"),
```
Comment (rawmind0) [inline rancher2/resource_rancher2_global_dns_provider_test.go]: What is `global_role_id` argument??

File: rancher2/resource_rancher2_global_dns_provider_test.go [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+	"testing"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	managementClient "github.com/rancher/types/client/management/v3"
+)
+
+const (
+	testAccRancher2GlobalDNSProviderType = "rancher2_global_dns_config"
```
Comment (rawmind0) [inline rancher2/resource_rancher2_global_dns_provider_test.go]: testAccRancher2GlobalDNSProviderType should be `"rancher2_global_dns_provider"`

File: rancher2/resource_rancher2_global_dns_provider_test.go [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+var (
+	testAccRancher2GlobalDNSProviderConfig         string
+	testAccRancher2GlobalDNSProviderUpdateConfig   string
+	testAccRancher2GlobalDNSProviderRecreateConfig string
+)
+
+func init() {
+	testAccRancher2GlobalDNSProviderConfig = `
+resource "rancher2_global_dns_config" "foo" {
```
Comment (rawmind0) [inline rancher2/resource_rancher2_global_dns_provider_test.go]: Resource should be "rancher2_global_dns_provider"

**Review by rawmind0 — CHANGES_REQUESTED** [review]
@dahendel , thanks for the PR. 

Acceptance test are wrong described and failing. To be sure that acceptance tests are passing, please execute `TESTACC_EXPOSE_HOST_PORTS=true make testacc` .Please take a look to the changes requested.

As general feedback, please review full code due to few references to `global role binding` on error and info messages.

Also, could you please add docs and include them in the PR??

File: rancher2/schema_global_dns_provider.go (hunk 2 above)
Comment (dahendel) [reply rancher2/schema_global_dns_provider.go]: Thanks, I was trying to find this. I didn't see it in the TF docs.

**Comment by rawmind0** [comment]
Fix issue #228

**Comment by dahendel** [comment]
Working on why the Global DNS Provider tests are failing. Not sure what's going on.

**Comment by rawmind0** [comment]
CI log is reporting a broken link `http://127.0.0.1:4567/docs/providers/rancher2/r/string` 

```
...
2020-02-23 18:28:14 URL:http://127.0.0.1:4567/docs/providers/rancher2/r/token.html [41047/41047] -> "token.html.tmp.tmp" [1]
2020-02-23 18:28:14 URL:http://127.0.0.1:4567/docs/providers/rancher2/r/user.html [40324/40324] -> "user.html.tmp.tmp" [1]
http://127.0.0.1:4567/docs/providers/rancher2/r/string:
Remote file does not exist -- broken link!!!
Found 1 broken link.
http://127.0.0.1:4567/docs/providers/rancher2/r/string
FINISHED --2020-02-23 18:28:14--
Total wall clock time: 1.9s
Downloaded: 72 files, 2.9M in 0.04s (67.3 MB/s)
make[1]: *** [website-provider-test] Error 8
make[1]: Leaving directory `/home/travis/gopath/src/github.com/hashicorp/terraform-website'
make: *** [website-test] Error 2
The command "make website-test" exited with 2.
```

Could you please check for website files??

**Comment by dahendel** [comment]
@rawmind0 I have no clue where that string is coming from. Searching the erb file and it simply isn't there.

File: website/docs/r/globalDNSProvider.html.markdown [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Optional/Computed/ForceNew) The name of the Global DNS Provider (string)
+* `dns_provider` - (Required/ForceNew) The Global DNS Provider [route53, alidns, cloudflare] (string)
```
Comment (rawmind0) [inline website/docs/r/globalDNSProvider.html.markdown]: Wrong format

```
-* `dns_provider` - (Required/ForceNew) The Global DNS Provider [route53, alidns, cloudflare] (string)
+* `dns_provider` - (Required/ForceNew) The Global DNS Provider `route53, alidns, cloudflare` (string)
```

File: website/docs/r/globalDNSProvider.html.markdown [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+
+## Nested blocks
+
+### `route_53`
+
+#### Arguments
+
+* `access_key` - (Required) The AWS Access key (string)
+* `secret_key` - (Required) The AWS Secret key (string)
+* `zone_type` - (Required/ForceNew) The Route53 zone type [public, private] (string)
```
Comment (rawmind0) [inline website/docs/r/globalDNSProvider.html.markdown]: Wrong format

```
-* `zone_type` - (Required/ForceNew) The Route53 zone type [public, private] (string)
+* `zone_type` - (Required/ForceNew) The Route53 zone type `public, private` (string)
```

File: website/docs/r/globalDNSEntry.html.markdown [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+	project_ids = ["${data.rancher2_project.proj.id}"]
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `name` - (Optional/Computed/ForceNew) The name of the Global DNS Entry (string)
+* `fqdn` - (Required/ForceNew) The Global DNS Entry [route53, alidns, cloudflare] (string)
```
Comment (rawmind0) [inline website/docs/r/globalDNSEntry.html.markdown]: Wrong format

```
-* `fqdn` - (Required/ForceNew) The Global DNS Entry [route53, alidns, cloudflare] (string)
+* `fqdn` - (Required/ForceNew) The Global DNS Entry `route53, alidns, cloudflare` (string)
```

**Review by rawmind0 — CHANGES_REQUESTED** [review]
Wrong format on doc files

File: rancher2/data_source_rancher2_global_dns_provider.go [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+)
+
+func dataSourceRancher2GlobalDNSProvider() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceRancher2GlobalDNSProviderRead,
+
+		Schema: map[string]*schema.Schema{
```
Comment (rawmind0) [inline rancher2/data_source_rancher2_global_dns_provider.go:14]: Datasource schema is incomplete. If using `flattenGlobalDNSProvider` schema should fit rancher2_global_dns_provider resource but computed fields.

**Review by rawmind0 — CHANGES_REQUESTED** [review]
Acceptance tests are failing on all 3 new objects, 2 resources and 1 datasource. Please, check that acceptance test are passing before submit.

```
...
=== RUN   TestAccRancher2GlobalDNSProviderDataSource
--- FAIL: TestAccRancher2GlobalDNSProviderDataSource (0.36s)
    testing.go:569: Step 0 error: errors during apply:

        Error: Unknown schema type [globalDnsProvider]

          on /var/folders/67/03ztmxrj7_q_km5qhq7qwz_40000gn/T/tf-test219552490/main.tf line 2:
          (source code not available)
...
=== RUN   TestAccRancher2GlobalDNSEntry_basic
--- FAIL: TestAccRancher2GlobalDNSEntry_basic (0.05s)
    testing.go:569: Step 0 error: config is invalid: Invalid reference: A reference to a resource type must be followed by at least one attribute access, specifying the resource name.
=== RUN   TestAccRancher2GlobalDNSEntry_disappears
--- FAIL: TestAccRancher2GlobalDNSEntry_disappears (0.05s)
    testing.go:569: Step 0 error: config is invalid: Invalid reference: A reference to a resource type must be followed by at least one attribute access, specifying the resource name.
=== RUN   TestAccRancher2GlobalDNSProvider_basic
--- FAIL: TestAccRancher2GlobalDNSProvider_basic (0.25s)
    testing.go:569: Step 0 error: errors during apply:

        Error: Unknown schema type [globalDnsProvider]

          on /var/folders/67/03ztmxrj7_q_km5qhq7qwz_40000gn/T/tf-test662525622/main.tf line 2:
          (source code not available)
=== RUN   TestAccRancher2GlobalDNSProvider_disappears
--- FAIL: TestAccRancher2GlobalDNSProvider_disappears (0.22s)
    testing.go:569: Step 0 error: errors during apply:

        Error: Unknown schema type [globalDnsProvider]

          on /var/folders/67/03ztmxrj7_q_km5qhq7qwz_40000gn/T/tf-test053965144/main.tf line 2:
          (source code not available)
```

**Comment by lots0logs** [comment]
@dahendel @rawmind0 :innocent: 

bump

**Comment by dahendel** [comment]
@rawmind0 any update on getting tests to pass when hitting a non-ha rancher?

@lots0logs and anyone else who finds this, the tests are failing because the rancher instance that is used for the test suite is a single node instance and in that install scenario the Global DNS Provider resources are not supported.

**Comment by lots0logs** [comment]
I see. Though it's technically just non-ha single node setups that aren't supported (since you can do a single node HA setup) right? I started to look at fixing up the branch the other day but got pulled into other stuff. As I understand it, the blockers are the tests passing and the issue mentioned [here]( https://github.com/terraform-providers/terraform-provider-rancher2/pull/253#discussion_r400356638). Is that right @rawmind0?

**Comment by dahendel** [comment]
@lots0logs This is the [issue]( https://github.com/rancher/rancher/issues/24962 )
We have a thread in the #terraform-provider-rancher2 slack channel


image

**Comment by rawmind0** [comment]
> @rawmind0 any update on getting tests to pass when hitting a non-ha rancher?

This just can be possible downgrading rancher server, but not considering this way. Moreover, WIP is intended to deploy rancher HA on acceptance tests.

@lots0logs, there are 2 blockers: implement changes requested and pass acceptance tests

**Comment by rawmind0** [comment]
@dahendel , HA acceptance tests are already merged on master. Could you please rebase this PR and try to pass them??

**Comment by dahendel** [comment]
@rawmind0 will do.

File: rancher2/data_source_rancher2_global_dns_provider.go (hunk 15 above)
Comment (dahendel) [reply rancher2/data_source_rancher2_global_dns_provider.go:14]: I updated this to what I think you meant. Can you verify?

**Comment by dahendel** [comment]
I rebased, but acceptance tests were failing outside of my code. Also, I saw it is commented out in the travis file.

**Comment by rawmind0** [comment]
> I rebased, but acceptance tests were failing outside of my code. Also, I saw it is commented out in the travis file.

@dahendel , acceptance tests should pass before we merge any PR. They're commented on travis file due to some duration issue on travis pipeline, but they should be passed manually before merge.
Acceptance tests are passing fine on master branch. How are you executing them?? Error log?? If you are trying to pass them on osx or windows, you should use `make docker-testacc` command.

**Comment by rawmind0** [comment]
@dahendel , any update here??

**Comment by dahendel** [comment]
@rawmind0 I rebased on master and having to fix a few things. Working on it as I can outside of work hours.

**Comment by abhi1693** [comment]
@rawmind0 Any ETA on this feature?

**Comment by rawmind0** [comment]
@dahendel , thanks for the PR, but we should move on with this feature. 

This PR is superseded by #551

**Pull request closed without merging**